keps/sig-apps/3329-non-retriable-exit-codes/README.md

```diff
@@ -0,0 +1,1008 @@
+such case, restarting such a pod results in wasting a lot of computational time.
+
+We would like to be able to automatically detect and terminate the jobs which are
+failing due to the bugs in code of the executable, so that the computation resources
+can be saved.
+
+#### Story 2
+
+As a service provider that offers computational resources to researchers I would like to
+have a mechanism which terminates jobs for which pods are failing due to user errors.
+It is important that the mechanism is can automatically detect such jobs as I do not
+have expertise in what the computations do. This is important in order to limit the
+costs of running the service.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!-- Increased complexity of the Job management, by interaction with some of
+already exiting Job configuration options. To mitigate this, we introduce
+minimal API which will allow to satisfy the main usage scenario.-->
+
+Container exit codes could be lost if the failed pod is garbage collected.
+First, it should be rather rare so the overall consequence of unnecessary pod restarts
+will be limited. Second, we can prevent this by using the feature of
+[job tracking with finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/).
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+
+### JobSpec API
+
+To be discussed. Here are some variants:
+
+#### Variant 1:
+
+Single job-level field with range values. In this
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: main-job-container
+        image: job-image
+        command: ["./program"]
+  backoffLimit: 3
+  terminateOnExitCodes: 1-12,100-127
+```
+
+Pros:
+- simple API extension satisfying the main usecase
+
+Cons:
+- the API may not be very flexible to extend in the future
+- single range of exit codes for all containers
+
+#### Variant 2
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: main-job-container
+        image: job-image
+        command: ["./program"]
+      - name: monitoring-job-container
+        image: job-monitoring
+        command: ["./monitoring"]
+  backoffLimit: 3
+  backoffPolicy:
+    terminateOnConditions:
+    - exitCodes: 1-12,100-127
+      containerName: main-job-container
+```
+
+Here, we could make the `containerName` field optional. If not specified,
```

Comment:
Include the type with all the fields to be added, like here https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2307-job-tracking-without-lingering-pods#api-changes